bn-apps/ledger-sync/ledger-sync-service/src/main/kotlin/net/corda/businessnetworks/ledgersync/TransactionRecoveryFlow.kt

```diff
@@ -0,0 +1,58 @@
+package net.corda.businessnetworks.ledgersync
+
+import co.paralleluniverse.fibers.Suspendable
+import net.corda.core.crypto.SecureHash
+import net.corda.core.flows.FlowLogic
+import net.corda.core.flows.FlowSession
+import net.corda.core.flows.InitiatedBy
+import net.corda.core.flows.InitiatingFlow
+import net.corda.core.flows.ReceiveTransactionFlow
+import net.corda.core.flows.SendTransactionFlow
+import net.corda.core.identity.Party
+import net.corda.core.utilities.unwrap
+
+/**
+ * This flow allows for the recovery of transactions with IDs found previously.
+ */
+@InitiatingFlow
+class TransactionRecoveryFlow(
+        private val report: Map<Party, LedgerSyncFindings>
+) : FlowLogic<Unit>() {
+
+    @Suspendable
+    override fun call() {
+        report.toList().map { (counterParty, findings) ->
+            val session = initiateFlow(counterParty)
+            findings.recoverable().also {
+                session.send(it)
+            }.map {
+                subFlow(ReceiveTransactionFlow(session))
+            }
+        }.flatten().onEach {
+            it.verify(serviceHub)
+        }.also {
+            serviceHub.recordTransactions(it)
+        }
+    }
+
+    private fun LedgerSyncFindings.recoverable() = missingAtRequester - missingAtRequestee
```

Comment:
I wonder if we need to do this substraction? Transaction can't be reported as missing from both requested and requeste, can it?

Reply:
You're right. At least not if the report is generated by `RequestLedgersSyncFlow` but if the source for the report is a different on it might make sense?

Reply:
Can't come up with such usecase to be honest. Happy to keep it as long it won't make things wrong :)